stdlib/public/runtime/SwiftDtoa.c

```diff
@@ -0,0 +1,2429 @@
+///   between two adjacent binary floating-point values. Theorem 6 of
+///   the Errol paper characterizes this sufficiently to allow us to
+///   vary our strategy. Errol3 uses a separate formatter for this
+///   case. This implementation instead widens the interval (as in
+///   Grisu3), which allows us to use the same fast digit generation in
+///   all cases, providing uniform performance across the entire range.
+///
+/// In addition to addressing the shortness failures characterized in
+/// the Errol paper, the implementation here also incorporates
+/// final-digit corrections from Grisu3, allowing it to produce the
+/// optimal decimal decomposition in all cases.
+///
+/// In summary, this implementation is:
+///
+/// * Fast.  It uses only fixed-width integer arithmetic and has
+///   constant memory requirements.
+///
+/// * Simple. It is only a little more complex than Loitsch' original
+///   implementation of Grisu2.  The full digit decomposition for double
+///   is less than 300 lines of standard C.
+///
+/// * Always Accurate. Converting the decimal form back to binary
+///   will always yield exactly the same value.  For the IEEE 754
+///   formats, the round-trip will produce exactly the same bit
+///   pattern in memory.
+///
+/// * Always Short.  This always selects an accurate result with the
+///   minimum number of decimal digits.
+///
+/// * Always Close.  Among all accurate, short results, this always
+///   chooses the result that is closest to the exact floating-point
+///   value. (In case of an exact tie, it rounds the last digit even.)
+///
+/// For single-precision Float, these claims have been exhaustively
+/// tested -- all 2^32 values can be checked in under an hour on a
+/// mid-range modern laptop. The Double and Float80 formatters rely on
+/// results from the Errol paper to ensure correctness.  In addition,
+/// we have verified more than 10^13 randomly-chosen values by comparing
+/// the results to the output of Errol4.
+///
+// ----------------------------------------------------------------------------
+
+#include <float.h>
+#include <math.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "swift/Runtime/SwiftDtoa.h"
+
+#if defined(__x86_64__) || defined(__aarch64__)
+  // We get a significant speed boost if we can use the __uint128_t
+  // type that's present in GCC and Clang on 64-bit architectures.  In
+  // particular, we can do 128-bit arithmetic directly and can
+  // represent 192-bit integers as a collection of 64-bit elements.
+  #define HAVE_UINT128_T 1
+#else
+  // On 32-bit, we have to use slower code that manipulates 128-bit
+  // and 192-bit integers as collections of 32-bit elements.
+  #define HAVE_UINT128_T 0
+#endif
+
+// Try to verify that the system floating-point types really are what we
+// expect.  Note that the code below is specific to these exact
+// floating-point representations.
```

Comment:
These are mostly intended as simple guards in case someone copies this code elsewhere.  However, even within the Swift stdlib, we cannot safely assume that C `long double` really is Float80, so that check at least is important to have here.

Reply:
Yup, that's fine.